x/multisig/decorator_test.go

```diff
@@ -0,0 +1,104 @@
+package multisig
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/store"
+	"github.com/iov-one/weave/x"
+)
+
+func TestDecorator(t *testing.T) {
+	var helpers x.TestHelpers
+
+	_, a := helpers.MakeKey()
+	_, b := helpers.MakeKey()
+	_, c := helpers.MakeKey()
+	auth := helpers.CtxAuth("multisig")
+	bg := auth.SetConditions(context.Background(), a, b, c)
+
+	h := new(MultisigCheckHandler)
+	d := NewDecorator(auth, NewContractBucket())
+	stack := helpers.Wrap(d, h)
+
+	db := store.MemStore()
+	hc := CreateContractMsgHandler{auth, NewContractBucket()}
+	res, err := hc.Deliver(bg, db,
+		helpers.MockTx(
+			&CreateContractMsg{
+				Sigs:                [][]byte{a.Address(), b.Address(), c.Address()},
+				ActivationThreshold: 2,
+				AdminThreshold:      3,
+			}))
+	require.NoError(t, err)
+	contractID := res.Data
+
+	multisigTx := func(payload, multisig []byte) ContractTx {
+		tx := helpers.MockTx(helpers.MockMsg(payload))
+		return ContractTx{Tx: tx, MultisigID: multisig}
+	}
+
+	cases := []struct {
+		tx    weave.Tx
+		perms []weave.Condition
+	}{
+		// doesn't support multisig interface
+		{helpers.MockTx(helpers.MockMsg([]byte{1, 2, 3})), nil},
+		// Correct interface but no content
+		{multisigTx([]byte("john"), nil), nil},
+		// with multisig contract
+		{multisigTx([]byte("foo"), contractID),
```

Comment:
looks good.
please add two more cases:
* invalid contractId -> error
* contractId but insufficient signatures -> error

Reply:
btw, have you ever looked at code coverage output? it is pretty nice.

try `make cover` then look at the html file for your package